### ISSUE TITLE
Future of Hometown

### ISSUE THREAD

**Issue by graue** [hometown-fork/hometown#1358]
For the past year, with the exception of one critical security vulnerability, Hometown seems to have received no maintenance. Our fork creator Darius Kazemi no longer seems to have time to maintain the fork due to other commitments. That's OK, it's perfectly natural for people to move on—Darius doesn't owe us indefinite maintenance—but it also appears no one else can merge PRs or issue a release.

To illustrate the problem, Mastodon 4.2 [came out](https://mastodon.social/@MastodonEngineering/111104350824645579) in late September 2023, a community PR [syncing Hometown with 4.2](https://github.com/hometown-fork/hometown/pull/1337) was deployed in production and ready for review by late December 2023, but despite multiple comments by Darius promising to get to reviewing the PR soon, no movement occurred until early July 2024, 9 months after 4.2 was released and 6 months after the PR was largely complete, when a critical security update forced Hometown to finally [release a 4.2 version](https://github.com/hometown-fork/hometown/releases/tag/v4.2.10%2Bhometown-1.1.1).

No maintenance has been done since that release. A [PR submitted one day later](https://github.com/hometown-fork/hometown/pull/1351) fixing a bug in that release has not been merged or reviewed after 3 months. Issues have not been acknowledged, and with Mastodon 4.3 coming out today with some major improvements, I don't have confidence Hometown will sync with 4.3 anytime soon.

So I wanted to open a space to talk with other Hometown admins on paths forward.

The options I see are: if someone is willing to step up and be a maintainer, they could take over the project either with Darius' blessing, or without (fork the fork). Or we could migrate to other server software, such as the Glitch fork which also has local-only posts; I'm not sure how difficult that would be. In my case, the server I co-admin has never used local-only posts very much, because we didn't attract enough critical mass of our community to our server for it to be useful, so we have considered just deleting the few local-only posts and switching to stock Mastodon. What are you thinking, Hometown admins?

**Comment by a-dows** [comment]
My three hometown servers have all grown happily for years without paying
much mind to mastodon vanilla’s feature releases. Hometown has been stable
and meets the asks of my users. I don’t need it to do anything it doesn’t
do already, as long as critical security updates continue (as they have).

**Comment by dariusk** [comment]
Hey all -- I would like for 2025 to be the year that Hometown changes governance structure so that I am no longer the sole maintainer/owner. Currently doing a bunch of bugfixes on Hometown in preparation for a Mastodon 4.3 merge and then I want to address this in earnest.

**Comment by rscmbbng** [comment]
As a longtime and happy Hometowner I'd like to push this conversation forward. It would be great to have some formal governance that can also raise some funds to pay folks to maintain and develop hometown. Would it be an idea to draft some scenarios together?

**Comment by 320x200** [comment]
Silly question from the happy-admin-of-a-hometown-instance department → has anyone, or a small group, already offered to take over the maintenance/ownership of the project yet? Alternatively, would you like to be involved @dariusk if this was a collective governance?

Regarding  the local post only question from @graue, at LURK we do not use it a lot. I think there was a lot of speculation and excitement about such a feature for theme/practice/topic based communities on small instance. But as it turns out, people tend to default to mass broadcasting... **However**. there are very specific situations in which we use local only (administrativia, semi-private stuff, jokes or more informal discussions, etc.) and it is essential even if uncommon and hard to predict when it makes sense to restrict the visibility to local users. I can't think of going back running an instance without such a feature.

**Comment by timcowlishaw** [comment]
Hey folks, just wanted to add my voice to this and hoping to revive the conversation - i'd be very happy to help out with maintenance / governance if @dariusk would find that useful.

We're in the process of moving assemblag.es from a managed hosting provider onto our own infrastructure in order to help secure the longer-term future of the community, and I see part of that as taking more responsibility for the software we're running too, so very happy to help out in any way i can.

Thanks!

Tim

**Comment by cblgh** [comment]
This sustainability issue was just raised on our instance - pinging @dariusk

What people / which groups are most able to take over after you Darius if nothing else were to change from the state of today? 

If people know this then they can reach out with further plans such as that mentioned by @rscmbbng and others above.

**Comment by nachtjasmin** [comment]
(edit: submitted a bit too early, sorry for the half-baked reply in the mails!)

I'm really curious to see what the future is upholding for Hometown. IMHO, but that's just my humble opinion, I think that the community would benefit from a **hard fork** of upstream Mastodon.

Why the hard fork, you might ask? Let me elaborate on that. As much as I'm grateful for the things that the Mastodon developer team achieved, I doubt that their alignment is helpful for the smaller instances out there. As much as quote toots and soon "packs" are features beneficial for users, I'm missing features that'd help us moderators and administrators to combat spam waves.

Anyway, because most of the features that were Hometown-exclusive are now in upstream Mastodon (except local-only posting and some other smaller things like the pastel themes), I'm not sure whether it makes sense to try to merge the current state of Hometown and Mastodon together. After all, the only feature that's critical (in the meaning of: ignoring it has serious consequences for previously local-only toots), it *might* make sense to start building that feature "from scratch" on top of Mastodon 4.3 (or 4.4).

I'm experimenting with exactly that approach right now and it is doable, albeit probably not the most elegant one. The primary challenge is spotting all changes where local-only posting is involved. You might be asking why I chose this approach? The idea is simple: With just one patch on top of upstream, it might allow us to upgrade to newer versions more easily.

**Comment by graue** [comment]
Making Hometown a hard fork doesn't make sense to me. It would be much more work to adequately maintain than to maintain the current soft fork, and Hometown already isn't attracting enough developer interest for that.

Much as I disagree with the upstream Mastodon project's priorities and choices, it's valuable to stay in sync enough to benefit from their continued UI improvement, feature development, and especially security updates. If I'm going to give all that up, then I'd rather switch entirely to another project like GoToSocial, which [has explored the possibility of a migration path](https://codeberg.org/superseriousbusiness/gotosocial/issues/128) from Mastodon.

**Comment by abochmann** [comment]
Commenting as someone wo runs a single-user Hometown instance, mainly for the UI improvements (direct links to original posts, better handling of long / formatted articles, easily configurable toot length, and yes, pastel theme) and additional privacy settings (no RSS feeds comes to mind, IDK which of the other admin options are in upstream by now).

Mastodon 4.2 will be supported only another two months (until 2026-01-08). I've had no problems applying upstream 4.2.x releases as a patch to Hometown up to now - but things will be more difficult after that date. In theory it's probably possible to backport new security fixes to 4.2 for some time if no one steps up to rebase Hometown, but seeing as either of that needs similar knowledge of the codebase, I don't have much hope for that to happen...

Would anyone be willing to work on a small migration guide for Hometown instances? I see two options:

- Back to upstream 4.2, and then upgrade from there, which probably involves dropping all existing local-only posts (or maybe convert them to posts visible only to mentioned accounts). Do Hometown's exclusive lists diverge from what Mastodon has implemented and would those need a migration? Anything else that needs to be cleaned up in the database or elsewhere?
- Over to Glitchsoc, which as far as I'm aware is the only other maintained fork that has local-only posts? Do they use a compatible implementation of those?

**Comment by nachtjasmin** [comment]
@abochmann The exclusive lists of Hometown and the one of upstream Mastodon are identical. The UI might differ, but they're using the same database scheme under the hood, as far as I remember. In terms of cleanup: From what I've seen, Hometown also stores the type of post (whether it's an `Article` or a `Note`) in the database, so that column would have to be removed (or kept, but it's useless after migrating away).

A migration to glitch-soc should be doable, I think. From what I've heard, their local-only implementation seems to be behind the Hometown implementation (in terms of usability), but that's just hearsay, so take it with a grain of salt. Code-wise, they should be identical or at least *close enough*, so that it's feasible to write a migration towards glitch-soc.

For a migration to upstream Mastodon however, it might be better to "just" delete all local-only posts or to "convert" them to direct messages. Information might get lost then, but it might be safer than an accidental leak of the local-only posts to other instances, in my opinion.

**Comment by cblgh** [comment]
just chiming in re

> For a migration to upstream Mastodon however, it might be better to "just" delete all local-only posts or to "convert" them to direct messages. Information might get lost then, but it might be safer than an accidental leak of the local-only posts to other instances, in my opinion.

i think this would prohibit a majority of instances from actually switching, if only for emotional reasons. i agree that from a technical standpoint it is 100% desirable to be certain local-only posts do not leak outside the instance. a possible mitigation for a migration path that does deletes local-only posts could be to couple it with some kind of archive tool à la https://tinysubversions.com/twitter-archive/make-your-own/ that @dariusk made for leaving twitter

with an archive tool in place, instances can present their community with a grace period where people can download all the local only posts + media into a single archival format that is actually useful for them

**Comment by biglifedecision** [comment]
I run a Hometown instance with ~100 active users. 

I am not interested in a hard-fork away from Mastodon. Although I might personally disagree with a lot of Mastodon's silly features, I selected Hometown for our community specifically because it intended to follow Mastodon 99.999% with local-only posts, which is the additional feature that is the "deal-breaker" for our community.

>Hometown is a light weight fork of Mastodon. By "light weight" I don't mean more efficient; I mean this fork is 99.999% identical to Mastodon with a few key tweaks. This project is based on the principle of: minimum code change for maximum user experience change. This makes it easy for the basically-one-person who runs the project to keep it up to date. By our best understanding, our major changes are not wanted by the Mastodon project, hence maintaining this fork instead of trying to commit the changes to Mastodon.

So, I feel if Hometown is to continue it should follow that ethos. For anyone wanting something different/more opinionated, projects like GoToSocial are probably a better fit? 

Currently my issue is less with missing out on Mastodon's newest shiny gimmicky features. Some of my users will like those but it isn't the priority. There are **a lot** of bug fixes, security fixes, and under-hood performance, maintenance, stability improvements that have happened to Mastodon that Hometown is lacking now. IMO this matters way more than the gimmick features Mastodon has added. 

I know of at least one larger instance that recently migrated from Hometown to the matching glitch-soc version, and then upgraded up to the most recent version. It went smoothly and my understanding is that local-only posts (the feature my community needs most) are working fine. This is the path I plan to take if Hometown is still stagnant at the end of the year. 

The other option for my needs (and maybe others?) is trying to revisit convincing Mastodon to upstream the local-only post feature. I think I could make a strong argument for why this feature would actually contribute to Mastodon's greater success, based on my experience. This was shot down in the past, but the Mastodon team is bigger and more diverse now. Perhaps it will be more open to this feature. If anyone can point me to the best place (issue in Mastodon's repo?) to raise this I would be happy to open it up and then direct people to support the proposal.

**Comment by rscmbbng** [comment]
I see discussion is veering towards "how to get out" but I'd first like to have a conversation about "how to stay in". 

There is merit to figuring out together how the work on Hometown can continue. The vision of Hometown is still very urgent: alternative social media geared towards smaller communities. Yet, the reality is that Hometown is outdated (some functionality is in main, or will be) and unmaintained (we are increasingly incompatible with other fedi software). However, vanilla Mastodon and Glitch (besides local-only) don't cater specifically to smaller communities. GoToSocial, while it is a very cool project, just does not have the same level of finish and maturity as Hometown/Mastodon. That is to say, there is space for something like Hometown. And there is a future for a fork that has (and continues to develop) crucial functionality for smaller communities while staying 99.999% the same as Mastodon.

I strongly believe there are sufficient resources and skills in the Hometown network to do this, and we need to put our heads together and figure out how. Immediate steps are to make a release that catches up with upstream, find a way to structurally pay someone to maintain it, and establish a form of non-BDFL governance. This may require a friendly soft fork for now, just to get the ball rolling [^1].  

I'd be happy to initiate a call with those of you who feel similar, @timcowlishaw @cblgh @graue @320x200 

[1^]: Darius has not been responsive in this thread for a year nor replied to my DMs on the matter. I don't think this is in bad faith, just a lot on the plate but also blocking. Such a friendly soft fork would reconcile when Darius has space to do so.

**Comment by cblgh** [comment]
Just noting that there is recent activity by Darius in the PR started by @mistydemeo: https://github.com/hometown-fork/hometown/pull/1371

As for what @rscmbbng wrote recently:

On the topic of bringing together people to steward Hometown itself, based on the repo's PR history @mistydemeo and @nachtjasmin seem to have been doing excellent work! As for myself I'm not a good candidate for a codebase like Mastodon's -- I'm participating in this thread as a delegate of the Merveilles instance to keep up to date on Hometown's future and understand the migration options :)

Agree on arriving at community options to pay for future Hometown's maintenance! Open Collective has been a very good option for many ad-hoc groups in the past

**Comment by biglifedecision** [comment]
@rscmbbng I agree 100%. I have just been following this Issue since it's creation/ tried to read the tea leaves of @dariusk's actions. I am exploring escape options because, as was mentioned by @abochmann , Mastodon 4.2 will be EOL after January. If Hometown is still stagnant by then with no signs of progress on this Issue, it seems prudent to consider migrating to other options, even if less ideal, at that point for the sake of my community and the stability/security of our instance. 

I am much more interested in staying on Hometown, if Hometown can be brought up to date with Mastodon, if there is renewed commitment from old or new maintainers, or new governance, or a soft fork with the same vision ("99.999%"), or etc. Please include me in a call or any other discussion you have off of here! I cannot contribute as a developer, but am happy to in other ways.

**Comment by mistydemeo** [comment]
Just for context on that PR: I run a Hometown server, and I'm working to merge in the latest Mastodon code so that we're fully up to date. I don't know that I'm the right person to be doing forward-looking feature development, but I'm happy to help contribute to merging in upstream Mastodon releases more often.

**Comment by dariusk** [comment]
Hi everyone. Sorry for not responding here for a year. I have been keeping track. Every time I sit down to write a response there is more to consider and it becomes even harder to respond.

So I will just fire off a message that is off the top of my head, and not particularly informed, and probably doesn't address the last year's worth of comments because they aren't all in my head at the moment. Apologies in advance.

### Work being done right now

I'm aware of the end-of-life for Mastodon 4.2 and you can track the work that mostly @mistydemeo is doing to incrementally upgrade us to v4.5, with some assistance from me, here: https://github.com/hometown-fork/hometown/pull/1371

The intent is to upgrade the underlying Mastodon version without new introducing feature work for Hometown (so like, there wouldn't be a toggle to disable Quote Posts server-wide, for example, since that's [not something Mastodon implemented](https://docs.joinmastodon.org/user/quote-posts/#faq-admin)).

### On forks and/or Hometown diverging more from upstream Mastodon
I have pretty strong opinions that it's good to keep Hometown basically on par with upstream Mastodon, for my own sanity as maintainer. I experimented with the alternative mobile UI back in an earlier Hometown version but the Mastodon team changes stuff so frequently that it was a nightmare to maintain. If people want to take it in a more hard-fork-ish direction, my inclination is to tell you to just hard fork Hometown itself and you have my blessing!

As it currently stands, if Mastodon upstream implemented local-only posts like they did with exclusive lists... I would be probably declare "mission accomplished" and shut down Hometown, put out a version upgrade that takes you 1:1 back to Mastodon upstream. I'd encourage people who want to fork it to a new project to do so. But that's probably not happening any time soon, despite Mastodon folks saying every now and then that they would like to implement local-only.

### My material situation
Just in the interest of being upfront, the material benefits I gain from being known as the Hometown maintainer are: 

1) [some percentage of $339/mo](https://www.patreon.com/tinysubversions) (at the time of this writing) that represents people supporting Hometown (but hard to distinguish from what my other projects bring to the table here)
2) reputation, which helps me do things like [get hired at Harvard](https://tinysubversions.com/notes/joining-asml/), be listened to when I participate in standards processes, and get invited to cool events that sometimes pay to fly me out. Though again: hard to tell what percentage of that is Hometown vs [the rest of my projects](https://tinysubversions.com/projects/). I have made a lot of stuff; not to brag, it's just a fact, and it all has cumulative effect on my reputation.
3) the utility of having Hometown and its features on my own server, Friend Camp. This is the main material benefit for me: it improves my social media life.

There are of course many non-material benefits; I truly feel warm and fuzzy knowing that so many small communities use Hometown. But part of why I say all this is I am not emotionally or financially or reputationally biased toward Hometown continuing as a me-project, or Hometown continuing at all, or to Hometown having forked projects that work in the same space. It's ultimately all fine by me. If the project went in a direction I didn't like, I would just continue maintaining my Friend Camp specific fork for my own community.

### On my absentee-ism
Not much to say here other than I'm sorry. The last two years have been personally very tumultuous and difficult for me; the only takeaway there really is that it's a failure case of the benevolent dictator model.

### Community governance
I would love to hand over governance to a community group.

I do not know how to explore this because I do not know who the main people with bandwidth to potentially take over project governance are. I do know that if Hometown became a community governed project, I would not want to be involved, mostly because I spend all my community-governance energy on W3C standards stuff. I know how much hard work it is. Much harder than being a benevolent if absentee dictator.

Another option would be for there to be a benevolent co-dicatorship. This would at least mitigate the problem where "Darius is having a rough time" doesn't affect the project so much.

### In conclusion,,,

Anyway those are some scattered thoughts. I plan to keep up with the thread from here; please feel free to let me know what above concerns I've certainly missed, or to introduce new concerns to the thread.

**Comment by evanp** [comment]
@dariusk thanks for the thoughtful approach! It's good to see such an important part of the Fediverse getting careful attention from its developer and community.

**Comment by biglifedecision** [comment]
Thank you @dariusk for the update! I feel very reassured. I'm happy to stay on 99.999% Mastodon + local posts with you as benevolent dictator maintainer. A Co-Dictator sounds like a nice middle ground.

As I mentioned I've also been thinking about campaigning Mastodon to adopt local-only posting. I think I can personally make a strong case... I have been thinking this would be more likely as the Mastodon team has expanded over recent years. With the most recent news it seems more likely than ever. Knowing that this would been by you as a positive conclusion to Hometown is encouraging for that as well.

**Comment by mistydemeo** [comment]
As an update to Darius's comment about a "benevolent co-dicatorship", there are now two additional maintainers with commit access: @nachtjasmin and myself. With three people now able to merge things and cut releases, my hope is that we'll be in a position to keep more thoroughly on top of new releases.
